File: src/utils/validation.ts
/**
 * Validation utilities
 */

export function isValidNumber(value: unknown): value is number {
  if (typeof value === 'number') return !Number.isNaN(value);
  if (typeof value === 'string' && value.trim() !== '') {
    const num = Number(value);
    return !Number.isNaN(num);
  }
  return false;
}

export function parseNumber(value: unknown): number | undefined {
  if (!isValidNumber(value)) return undefined;
  return typeof value === 'number' ? value : Number(value);
}

export function isValidUrl(string: string): boolean {
  try {
    const url = new URL(string);
    return url.protocol === 'http:' || url.protocol === 'https:';
  } catch {
    return false;
  }
}

export function parseStringList(value: unknown): string[] | undefined {
  if (typeof value !== 'string') return undefined;
  return value
    .split(',')
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
}

export function isValidSearchType(value: unknown): value is string {
  const validTypes = [
    'auto',
    'fast',
    'deep',
    'instant',
    'neural',
    'keyword',
    'hybrid',
    'deep-lite',
    'deep-reasoning',
  ];
  return typeof value === 'string' && validTypes.includes(value);
}

export function isValidAnswerModel(value: unknown): value is string {
  const validModels = ['exa', 'exa-pro'];
  return typeof value === 'string' && validModels.includes(value);
}

export function isValidResearchModel(value: unknown): value is string {
  const validModels = ['fast', 'regular', 'pro'];
  return typeof value === 'string' && validModels.includes(value);
}

File: internal/utils/validation.go
package utils

import (
	"net/url"
	"strconv"
	"strings"
)

var validSearchTypes = map[string]bool{
	"auto": true, "neural": true, "keyword": true, "hybrid": true,
	"fast": true, "instant": true, "deep": true, "deep-lite": true,
	"deep-reasoning": true,
}

var validAnswerModels = map[string]bool{
	"exa": true, "exa-pro": true,
}

// ParseNumber parses a string as an integer. Returns 0, false if invalid.
func ParseNumber(s string) (int, bool) {
	s = strings.TrimSpace(s)
	if s == "" {
		return 0, false
	}
	n, err := strconv.Atoi(s)
	return n, err == nil
}

// ParseStringList splits a comma-separated string, trims whitespace, filters empties.
func ParseStringList(s string) []string {
	if s == "" {
		return nil
	}
	parts := strings.Split(s, ",")
	var result []string
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p != "" {
			result = append(result, p)
		}
	}
	return result
}

// IsValidURL returns true if s is a valid http or https URL.
func IsValidURL(s string) bool {
	u, err := url.Parse(s)
	if err != nil {
		return false
	}
	return u.Scheme == "http" || u.Scheme == "https"
}

// IsValidSearchType returns true if s is a recognized search type.
func IsValidSearchType(s string) bool {
	return validSearchTypes[s]
}

// IsValidAnswerModel returns true if s is a recognized answer model.
func IsValidAnswerModel(s string) bool {
	return validAnswerModels[s]
}
